perf_test/sparse/KokkosSparse_sptrsv_cholmod.cpp

```diff
@@ -0,0 +1,579 @@
+          // read CHOLMOD factor int crsMatrix on the host (cholmodMat_host) and copy to default host/device (cholmodMtx)
+          timer.reset();
+          std::cout << " > Read Cholmod factor into KokkosSparse::CrsMatrix (invert diagonabl, and copy to device) " << std::endl;
+          bool cusparse = true;
+          bool invert_diag = false;
+          auto graph = read_cholmod_graphL<graph_t>(cusparse, L, &cm);
+          auto cholmodMtx = read_cholmod_factor<crsmat_t, graph_t> (cusparse, invert_diag, L, &cm, graph);
+          std::cout << "   Conversion Time: " << timer.seconds() << std::endl << std::endl;
+
+          bool col_majorL = true;
+          bool col_majorU = false;
+          if (!check_cusparse(Mtx, col_majorL, cholmodMtx, col_majorU, cholmodMtx, perm, perm, tol, loop)) {
+            num_failed ++;
+          }
+        }
+        break;
+
+        default:
+          std::cout << " > Testing only Cholmod < " << std::endl;
+          exit(0);
+      }
+    }
+  }
+  std::cout << std::endl << std::endl;
+
+  return num_failed;
+}
+
+
+void print_help_sptrsv() {
+  printf("Options:\n");
+  printf("  --test [OPTION] : Use different kernel implementations\n");
+  printf("                    Options:\n");
+  printf("                    cholmod_naive, cholmod_etree\n\n");
+  printf("  -f [file]       : Read in Matrix Market formatted text file 'file'.\n");
+  printf("  --offset [O]    : Subtract O from every index.\n");
+  printf("                    Useful in case the matrix market file is not 0 based.\n\n");
+  printf("  -rpt [K]        : Number of Rows assigned to a thread.\n");
+  printf("  -ts [T]         : Number of threads per team.\n");
+  printf("  -vl [V]         : Vector-length (i.e. how many Cuda threads are a Kokkos 'thread').\n");
+  printf("  --loop [LOOP]   : How many spmv to run to aggregate average time. \n");
+}
+
+
+int main(int argc, char **argv)
+{
+  std::vector<int> tests;
+  std::string filename;
+
+  int loop = 1;
+
+  if(argc == 1)
+  {
+    print_help_sptrsv();
+    return 0;
+  }
+
+  for(int i=0;i<argc;i++)
+  {
+    if((strcmp(argv[i],"--test")==0)) {
+      i++;
+      if((strcmp(argv[i],"cholmod-naive")==0)) {
+        tests.push_back( SUPERNODAL_NAIVE );
+      }
+      if((strcmp(argv[i],"cholmod-etree")==0)) {
+        tests.push_back( SUPERNODAL_ETREE );
+      }
+      if((strcmp(argv[i],"cusparse")==0)) {
+        tests.push_back( CUSPARSE );
+      }
+      continue;
+    }
+    if((strcmp(argv[i],"-f")==0)) {
+      filename = argv[++i];
+      continue;
+    }
+    if((strcmp(argv[i],"--loop")==0)) {
+      loop = atoi(argv[++i]);
+      continue;
+    }
+    if((strcmp(argv[i],"--help")==0) || (strcmp(argv[i],"-h")==0)) {
+      print_help_sptrsv();
+      return 0;
+    }
+  }
+
+  std::cout << std::endl;
+  for (size_t i = 0; i < tests.size(); ++i) {
+    std::cout << "tests[" << i << "] = " << tests[i] << std::endl;
+  }
+
+  Kokkos::initialize(argc,argv);
```

Comment:
Please use `Kokkos::ScopeGuard` instead of `initialize` and `finalize`.  That's just one line of code instead of two, and `ScopeGuard` will ensure that `finalize` gets called even if an uncaught exception happens.

Reply:
We switched to using Kokkos::ScopeGuard!!